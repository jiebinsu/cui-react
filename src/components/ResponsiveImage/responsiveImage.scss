@import "@coopdigital/shared-component--image/dist/image.css";
