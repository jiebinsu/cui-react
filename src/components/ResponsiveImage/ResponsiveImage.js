import React from "react";
import PropTypes from "prop-types";
import uuid from "uuid";
import "./responsiveImage.scss";

const generateSourceMap = (sources) =>
  sources.map((source) => {
    const { type, media, srcSet } = source;
    const srcSets = Object.entries(srcSet).reduce(
      (acc, [descriptor, imgSrc]) => {
        acc.push(`${imgSrc} ${descriptor}`);
        return acc;
      },
      []
    );

    const sourceProps = { type, media };
    return <source key={uuid()} srcSet={srcSets.join(", ")} {...sourceProps} />;
  });

const ResponsiveImage = ({ src, alt, width, height, sources }) => {
  return (
    <figure className="coop-c-image">
      <picture className="coop-c-image__media">
        {generateSourceMap(sources)}
        <img src={src} alt={alt} width={width} height={height} />
      </picture>
    </figure>
  );
};

ResponsiveImage.defaultProps = {
  width: null,
  height: null,
  sources: [],
};

ResponsiveImage.propTypes = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string.isRequired,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  sources: PropTypes.arrayOf(PropTypes.object),
};

export default ResponsiveImage;
