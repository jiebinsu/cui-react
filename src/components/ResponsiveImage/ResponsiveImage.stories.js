import React from "react";
import ResponsiveImage from "./ResponsiveImage";

export default {
  title: "Base/ResponsiveImage",
  component: ResponsiveImage,
};

const src =
  "https://images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=png&q=60&w=600&h=338";
const alt = "How Co-op is supporting local communities";
export const Basic = (args) => <ResponsiveImage {...args} />;
Basic.args = {
  src,
  alt,
};

const sourcesTest = [
  {
    type: "image/webp",
    media: "(min-width: 1200)",
    srcSet: {
      "1x":
        "//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338",
      "2x":
        "//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338",
    },
  },
  {
    type: "image/webp",
    media: "(min-width: 600)",
    srcSet: {
      "1x":
        "//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338",
      "2x":
        "//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338",
    },
  },
];
export const WithSources = () => (
  <ResponsiveImage src={src} alt={alt} sources={sourcesTest} />
);
