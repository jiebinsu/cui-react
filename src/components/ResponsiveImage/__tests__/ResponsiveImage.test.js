import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import ResponsiveImage from "../index";

describe("ResponsiveImage", () => {
  it("should render correctly", () => {
    const wrapper = mount(<ResponsiveImage src="src" alt="alt" />);
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer
      .create(<ResponsiveImage src="src" alt="alt" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should allow width and height attributes on default src", () => {
    const tree = renderer
      .create(<ResponsiveImage src="src" alt="alt" width="120" height="120" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render sources", () => {
    const sourcesTest = [
      {
        type: "image/webp",
        media: "(min-width: 1200)",
        srcSet: {
          "1x":
            "//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338",
          "2x":
            "//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338",
        },
      },
      {
        type: "image/webp",
        media: "(min-width: 600)",
        srcSet: {
          "1x":
            "//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338",
          "2x":
            "//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338",
        },
      },
    ];
    const tree = renderer
      .create(<ResponsiveImage src="src" alt="alt" sources={sourcesTest} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
