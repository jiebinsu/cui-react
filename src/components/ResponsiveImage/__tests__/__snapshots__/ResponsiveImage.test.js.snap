// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsiveImage should allow width and height attributes on default src 1`] = `
<figure
  className="coop-c-image"
>
  <picture
    className="coop-c-image__media"
  >
    <img
      alt="alt"
      height="120"
      src="src"
      width="120"
    />
  </picture>
</figure>
`;

exports[`ResponsiveImage should render correctly 1`] = `
<figure
  className="coop-c-image"
>
  <picture
    className="coop-c-image__media"
  >
    <img
      alt="alt"
      height={null}
      src="src"
      width={null}
    />
  </picture>
</figure>
`;

exports[`ResponsiveImage should render sources 1`] = `
<figure
  className="coop-c-image"
>
  <picture
    className="coop-c-image__media"
  >
    <source
      media="(min-width: 1200)"
      srcSet="//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338 1x, //images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338 2x"
      type="image/webp"
    />
    <source
      media="(min-width: 600)"
      srcSet="//images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338 1x, //images.ctfassets.net/bffxiku554r1/7XDwxNP8ioupeMdJZxNS8/d23d2b802d6be0ecbcb7abd1ff33b159/coop-supporting-local-communities.png?fm=webp&amp;q=60&amp;w=600&amp;h=338 2x"
      type="image/webp"
    />
    <img
      alt="alt"
      height={null}
      src="src"
      width={null}
    />
  </picture>
</figure>
`;
